### PR TITLE
Add clamp_below and clamp_above functions.

### PR DESCRIPTION
**Pull request by sumerman**
They allow pruning an istore beyond a specified threshold and set
a sum of pruned values at the threshold key.

```SQL
SELECT clamp_below('-2=>1, -1=>1, 1=>1'::istore, 0);
    clamp_below
--------------------
 "0"=>"2", "1"=>"1"
(1 row)

SELECT clamp_above('-1=>1, 1=>1, 2=>1'::istore, 0);
     clamp_above
---------------------
 "-1"=>"1", "0"=>"2"
(1 row)
```

### CODE REVIEW

**Comment by sumerman** [comment]
Note: don't merge this just yet — I want to ponder if there is a way to get rid of unnecessary code duplication.

**Comment by sumerman** [comment]
With the new implementation it's easy to implement a single function that would perform either or both clamps:
```SQL
SELECT clamp('-2=>1, -1=>1, 1=>1'::istore, 0, NULL);
    clamp
--------------------
 "0"=>"2", "1"=>"1"
(1 row)

SELECT clamp('-1=>1, 1=>1, 2=>1'::istore, NULL, 0);
     clamp
---------------------
 "-1"=>"1", "0"=>"2"
(1 row)

SELECT clamp('-1=>1, 1=>1, 2=>1'::istore, 0, 0);
     clamp
---------------------
 "0"=>"3"
(1 row)
```

I think it'd be a better interface, wouldn't it?

File: src/bigistore.c [hunk 1]
```diff
@@ -1099,6 +1099,58 @@ Datum bigistore_slice_to_array(PG_FUNCTION_ARGS)
...
 
+static void bigistore_clamp_pass(BigIStore * is, int32 end_key, int delta_dir)
+{
+    BigIStorePair * pairs;
+    int64 result = 0;
+    int   index  = 0, count = 0;
+
+    pairs = FIRST_PAIR(is, BigIStorePair);
+    index = delta_dir > 0 ? 0 : is->len - 1;
+    while ( ((delta_dir > 0) && (index < is->len && pairs[index].key <= end_key)) ||
```
Comment (rapimo) [inline src/bigistore.c]: it's not wrong but excluding the equality case would save a loop call
and avoid setting the index back later

File: src/bigistore.c [hunk 2]
```diff
@@ -1099,6 +1099,58 @@ Datum bigistore_slice_to_array(PG_FUNCTION_ARGS)
...
+            ((delta_dir < 0) && (index >= 0      && pairs[index].key >= end_key)) )
+    {
+        is->buflen -= bigis_pair_buf_len(pairs + index);
+        result = DirectFunctionCall2(int8pl, result, pairs[index].val);
+        index += delta_dir, count++;
+    }
+
+    if (count) {
```
Comment (rapimo) [inline src/bigistore.c]: best practice / clarity  `if (count > 0)`

File: src/bigistore.c [hunk 3]
```diff
@@ -1099,6 +1099,58 @@ Datum bigistore_slice_to_array(PG_FUNCTION_ARGS)
...
+    BigIStore * is      = PG_GETARG_BIGIS_COPY(0);
+    int32       end_key = PG_GETARG_INT32(1);
+    bigistore_clamp_pass(is, end_key, 1);
+    PG_RETURN_POINTER(is);
+}
+
+PG_FUNCTION_INFO_V1(bigistore_clamp_above);
+Datum bigistore_clamp_above(PG_FUNCTION_ARGS)
+{
+    BigIStore * is      = PG_GETARG_BIGIS_COPY(0);
```
Comment (rapimo) [inline src/bigistore.c]: This implementation optimizes for the clamp case. Maybe it's worth checking the performance benefit / penalty 
of not getting a copy and only creating a new istore in the case there is actually something to clamp

Comment (sumerman) [reply src/bigistore.c]: Very good point.

File: src/bigistore.c (hunk 1 above)
Comment (sumerman) [reply src/bigistore.c]: If `key == end_key` is present in the `istore` we actually need both, add it to the result and decrease `buflen`.

Comment (rapimo) [reply src/bigistore.c]: yep but you can do it later in the if(count) i.e. `clamp_below('0=>2',0)` would run into unnecessary code

Comment (sumerman) [reply src/bigistore.c]: I'd rather leave the loop alone, and not call into `clamp_pass` at all, as you have suggested in another thread

File: src/istore.c [hunk 4]
```diff
@@ -1096,6 +1096,58 @@ Datum istore_slice_to_array(PG_FUNCTION_ARGS)
...
+    int32 result = 0;
+    int   index  = 0, count = 0;
+
+    pairs = FIRST_PAIR(is, IStorePair);
+    index = delta_dir > 0 ? 0 : is->len - 1;
+    while ( ((delta_dir > 0) && (index < is->len && pairs[index].key <= end_key)) ||
+            ((delta_dir < 0) && (index >= 0      && pairs[index].key >= end_key)) )
+    {
+        is->buflen -= is_pair_buf_len(pairs + index);
+        result = DirectFunctionCall2(int4pl, result, pairs[index].val);
```
Comment (rapimo) [inline src/istore.c]: `int4pl` it rather short all it does is overflow checking and I guess that a `DirectFunctionCall2` is potentially expensive. Maybe it's worth using [INTPL](https://github.com/adjust/istore/blob/master/src/istore.h#L205)

Comment (sumerman) [reply src/istore.c]: The only other place I found that uses this macro is `(big)istore_avl_transfn` everything else e.g. `istore_accumulate` uses `DirectFunctionCall2` (that's why I also went this way). I don't have an opinion on the matter, but I'm curious why it was written that way.

Comment (rapimo) [reply src/istore.c]: yep this performance optimization was introduced with 8cc0510bc63ecee11ad8772ace1eb392446377e5

Comment (rapimo) [reply src/istore.c]: the `DirectFunctionCall2` was used to dryout code through a functional approach in `istore_apply_datum`

Comment (sumerman) [reply src/istore.c]: Thanks for the explanation.

File: src/bigistore.c [hunk 5]
```diff
@@ -1136,18 +1136,26 @@ static void bigistore_clamp_pass(BigIStore * is, int32 end_key, int delta_dir)
...
 
 PG_FUNCTION_INFO_V1(bigistore_clamp_above);
 Datum bigistore_clamp_above(PG_FUNCTION_ARGS)
 {
-    BigIStore * is      = PG_GETARG_BIGIS_COPY(0);
-    int32       end_key = PG_GETARG_INT32(1);
-    bigistore_clamp_pass(is, end_key, -1);
+    BigIStore * is = PG_GETARG_BIGIS(0);
+    int32 end_key  = PG_GETARG_INT32(1);
+    if( (FIRST_PAIR(is, BigIStorePair))[is->len - 1].key > end_key )
```
Comment (rapimo) [inline src/bigistore.c]: how about introducing `LAST_PAIR` ?

**Comment by rapimo** [comment]
for me ec656ea now looks like a lazy refactoring I'd rather rewrite clamp_pass to deal with the fact that it has to create a new istore and instead of memmove do an appropriate memcpy i.e. set the first pair and copy the rest / copy the first n pair and set the last

File: src/bigistore.c [hunk 6]
```diff
@@ -1099,6 +1099,65 @@ Datum bigistore_slice_to_array(PG_FUNCTION_ARGS)
...
+
+    /* copy survivors into a new spot */
+    if (delta_dir > 0)
+        pairs = pairs + index;
+
+    new_size       = is->len - count;
+    bytesize       = ISHDRSZ + new_size * sizeof(BigIStorePair);
+    result_is      = palloc0(bytesize);
+    result_is->len = new_size;
+
```
Comment (rapimo) [inline src/bigistore.c]: that looks like `FINALIZE_ISTORE` / `FINALIZE_BIGISTORE`

Comment (sumerman) [reply src/bigistore.c]: Only this doesn't depend on having intermediate pairs structure and array allocated.

Comment (sumerman) [reply src/bigistore.c]: Can wrap the existing pairs into that struct though...

Comment (sumerman) [reply src/bigistore.c]: `FINALIZE_ISTORE` tries to `pfree` `pairs` in the end.

Comment (sumerman) [reply src/bigistore.c]: I introduced an intermediate `FINALIZE_ISTORE_BASE` in the next commit that doesn't `pfree`

Comment (rapimo) [reply src/bigistore.c]: what's wrong with the `pfree` ?

Comment (rapimo) [reply src/bigistore.c]: nvm a saw it already

**Comment by sumerman** [comment]
🎉 thank you!

**Pull request merged**